### PR TITLE
Support hashes for struct initializers

### LINKED ISSUE #3120: Ruby: Allow for the instantiation of nested message fields with ruby-hashes
# What

Currently instantiation of a protobuf message can be done using a ruby-hash but does not support instantiating the nested messages using nested hashes. Being able to support the instantiation of messages including nested messages using a single hash would be a **big** usability win for [googleapis/toolkit](https://github.com/googleapis/toolkit).

# Example
### Current
```rb
message = MyMessage.new({
  submessage_field: SubMessage.new(:foo => 42)
})
```
### Requested
```rb
message = MyMessage.new({
  submessage_field: {foo: 42}
})
```

### PR DESCRIPTION
**Pull request by jbolinger**
This change simplifies working with `Google::Protobuf::Struct` and `Google::Protobuf::Value` so that code like this:

```ruby
Google::Protobuf::Struct.new(fields: {
  'a' => Google::Protobuf::Value.new({number_value: 2.0})
})
```
can be written as:

```ruby
Google::Protobuf::Struct.new(fields: {
   'a' => {number_value: 2.0}
})
```

This would simplify struct usage in many practical cases, such as: https://github.com/googleapis/google-cloud-ruby/issues/2839

I believe this also resolves #3120.

### CODE REVIEW

File: ruby/tests/well_known_types_test.rb [hunk 1]
```diff
@@ -139,4 +139,31 @@ def test_any
     assert any.is(Google::Protobuf::Timestamp)
     assert_equal ts, any.unpack(Google::Protobuf::Timestamp)
   end
+
+  def test_struct_assign
+    s = Google::Protobuf::Struct.new(fields: {'a' => Google::Protobuf::Value.new({number_value: 4.4})})
+    assert_equal 4.4, s['a']
+
+    s = Google::Protobuf::Struct.new(fields: {'a' => {number_value: 2.2}})
```
Comment (blowmage) [inline ruby/tests/well_known_types_test.rb]: i would love to see the `fields` Hash keys be converted to Strings, so we could also use this:

```ruby
s = Google::Protobuf::Struct.new(fields: {a: {number_value: 2.2}})
```

File: ruby/tests/well_known_types_test.rb [hunk 2]
```diff
@@ -139,4 +139,31 @@ def test_any
...
+    assert_equal 2.2, s['a']
```
Comment (blowmage) [inline ruby/tests/well_known_types_test.rb]: i would also like to see `Google::Protobuf::Struct#[]` convert the method argument to a String, so this would also work:

```ruby
assert_equal 2.2, s[:a]
```

File: ruby/tests/well_known_types_test.rb (hunk 1 above)
Comment (jbolinger) [reply ruby/tests/well_known_types_test.rb]: you got it

File: ruby/tests/well_known_types_test.rb (hunk 2 above)
Comment (jbolinger) [reply ruby/tests/well_known_types_test.rb]: you got it. How's this look?

Comment (blowmage) [reply ruby/tests/well_known_types_test.rb]: Looks great!

File: ruby/tests/well_known_types_test.rb [hunk 3]
```diff
@@ -165,5 +168,17 @@ def test_struct_nested_assign
     assert_equal expected_b_x, s['b']['x'].values
     assert_equal true, s['b']['y']
```
Comment (blowmage) [inline ruby/tests/well_known_types_test.rb]: I would like to see these access checks made with string, symbols, and a combination, to make it really clear they are checking the same thing.

```ruby
assert_equal true, s['b']['y']
assert_equal true, s[:b][:y]
assert_equal true, s[:b]['y']
assert_equal true, s['b'][:y]
```

Comment (jbolinger) [reply ruby/tests/well_known_types_test.rb]: Makes sense - added.

File: ruby/ext/google/protobuf_c/map.c [hunk 4]
```diff
@@ -71,6 +71,9 @@ static VALUE table_key(Map* self, VALUE key,
     case UPB_TYPE_BYTES:
     case UPB_TYPE_STRING:
       // Strings: use string content directly.
+      if (TYPE(key) == T_SYMBOL) {
```
Comment (TeBoring) [inline ruby/ext/google/protobuf_c/map.c]: Why we have to this in order to support Struct?

Comment (jbolinger) [reply ruby/ext/google/protobuf_c/map.c]: I don't think so, and it didn't break any tests that I'm aware of. Can you think of a case where it would make a difference?

This would not throw type errors in some cases where it would have previously but I don't know if that should be considered additive vs a breaking change?

Comment (TeBoring) [reply ruby/ext/google/protobuf_c/map.c]: It doesn't break existing test cases. However, it allows symbol for string map, which is not allowed previously.

Comment (jbolinger) [reply ruby/ext/google/protobuf_c/map.c]: Right, is there any reason to prohibit that?

Comment (TeBoring) [reply ruby/ext/google/protobuf_c/map.c]: It makes map not consistent with optional, repeated fields.
Besides, how to compare symbol? identity compare? value compare? Is the semantic same as string?
Besides, is the feature quite important? Otherwise, we don't want to over-provide it, as these trivial features necessarily makes our library complicate.

Comment (TeBoring) [reply ruby/ext/google/protobuf_c/map.c]: It also breaks existing type checking. We should added tests for our type checking.

Comment (blowmage) [reply ruby/ext/google/protobuf_c/map.c]: It allows users to provide:

```ruby
Google::Protobuf::Struct.new(fields: {
   a: {number_value: 2.0}
})
```

instead of:

```ruby
Google::Protobuf::Struct.new(fields: {
   'a' => {number_value: 2.0}
})
```

I would argue that Converting `Symbol` key values to `String` in `Struct` is a beneficial affordance, as `Struct` will only every have `String` keys. `Map` can have any any value type as a key, so it wouldn't make as much sense.

That said, I do think it would be fine to accept `Symbol` values every time a `String` is expected. In Ruby, I would write it like this:

```ruby
class Google::Protobuf::Struct
  def [](key)
    inner_struct[String(key)]
  end
end
```

Basically, anytime we expect a `String` object, pass it to the `String()` method to perform the necessary conversions.

**Comment by jbolinger** [comment]
@TeBoring what kind of type checking tests did you have in mind? I'm happy to add them but I'm not sure exactly what you're asking for.

**Comment by TeBoring** [comment]
The check test I have in mind is something like:
```
// expect throw error
a.foo = 1  // foo is a string field
```
However, this could be a separate change.

**Comment by TeBoring** [comment]
If ruby users do expect Symbol to be the same as string, I think we should make it consistent in all our APIs.
However, we need a API review to allow that. @jbolinger Could you write a doc for that?
My suggestion is allowing Symbol could be another change. We don't need to do it in this change.

**Comment by jbolinger** [comment]
I'm not sure that the inconsistency is all that bad. Backing up a bit, the first version of this PR didn't include the symbol bit. I was trying to fix an existing issue that prevented using a hash to init a map value:

So, instead of this:

```ruby
Google::Protobuf::Struct.new(fields: {
  'a' => Google::Protobuf::Value.new({number_value: 2.0})
})
```

one could write a more concise version:

```ruby
Google::Protobuf::Struct.new(fields: {
   'a' => {number_value: 2.0}
})
```

I shouldn't speak for @blowmage, but he pointed out that getting rid of the hash rocket would look even better. It's shorter and likely the more commonly used syntax. So, I added the implicit string/symbol conversion to allow this:

```ruby
Google::Protobuf::Struct.new(fields: {
   a: {number_value: 2.0}
})
```

Sure, it's a little inconsistent but since a map key can't be symbol allowing either to be used in this context seems like a useful convenience. I'm not sure if I understand what you mean by making it consistent for optional fields or treating string/symbol as the same, etc. but I don't think it makes sense to support something like this:

```ruby
msg.a_string = :some_thing
```

That should probably require an explicit conversion and, even if we did support it, I'm not sure it would actually be useful in very many situations.

**Comment by jbolinger** [comment]
@TeBoring any updates on this (and could you run the tests again)?

**Comment by TeBoring** [comment]
I mean please make the Symbol a separate change.

**Comment by jbolinger** [comment]
@TeBoring what other use cases / changes did you have in mind? It looks like Symbol behaves as a string in certain contexts already (see my previous example). What else should we support?

**Comment by TeBoring** [comment]
I see your comment. So you only want to make string map key to accept
symbol?

**Comment by TeBoring** [comment]
Only for the map in struct?

**Comment by TeBoring** [comment]
If you could make the change only affect the map in struct but not other
maps, I am fine with that.

**Comment by TeBoring** [comment]
Never mind. I think I can accept this change. I want to let @haberman to also take a look, since this affects API.

**Comment by jbolinger** [comment]
Thanks @TeBoring! I see a build failure but I don't think it's related to this, but let me know if it is or if anything else needs to be done.

**Comment by jbolinger** [comment]
Are you still interested in accepting this change? I think it's a nice simplification but it fell off my radar. Let me know if I can do anything else.

**Comment by TeBoring** [comment]
It seems ruby 2.3 still failed?

**Comment by jbolinger** [comment]
Hmm, that's sort of a strange error that looks like it might be related to the rvm or something unrelated in the environment (openssl?). I can try to reproduce it when I'm back in the office on Monday.

**Comment by TeBoring** [comment]
Maybe sync and push?

**Comment by TeBoring** [comment]
https://www.ruby-lang.org/en/news/2019/03/31/support-of-ruby-2-3-has-ended/

Since the support for ruby 2.3 has been ended, we will deprecate ruby 2.3.
